### PR TITLE
[User Model] Rename notification event listeners

### PR DESCRIPTION
**Pull request by brismithers**
# Description
## One Line Summary
Rename notification event listeners based on new naming convention.

## Details
* Renamed all "handlers" to be either a "listener" or an "observer" depending on their purpose.  All callbacks now have pattern "onXXXXXX" for each callback.
* Make all notification listeners 0:* (some were 0:1)
* Use the terminology "click" instead of "open".
* Add IEventNotifier.hasSubscribers to determine if there are subscribers. A few implementing classes had to add this.
* Replaced `complete()` pattern with `preventDefault()` pattern in NSE notification received callback and foreground onWillDisplay callback.
* Add IDisplayableNotification and IDisplayableMutableNotification which adds a `display()` function to an INotification. This allows the manual display of a notification, when `event.preventDefault()` is called.
* Internally, rework NotificationGenerationProcessor to accommodate `preventDefault` pattern. The notification is now constructed when the notification job is constructed and is never copied/altered until the end of its lifecycle.
* Update example application for new naming conventions and patterns.
* Update for new naming conventions and patterns

### Motivation
A desire to align terminologies and naming conventions across the OneSignal SDKs.

### Scope
This change is primarily a name change within the `notifications` module/namespace.  There are some functional changes related to notification generation processing due to a programming pattern change (i.e. change from using `complete()` to using `preventDefault()`).

# Testing
## Unit testing
Existing unit tests related to notification generation were updated to new expected behavior.  Mostly this involved adding new mock calls that were previously not required.  

## Manual testing
Ran the example app with both a NSE and a foreground lifecycle listener, checking for different variations and manually inspecting behavior:

1. NSE calls `event.preventDefault()` and returns immediately -> Notification does not show
2. NSE calls `event.preventDefault()` then sleeps for 40 seconds before returning -> Notification shows after 30 seconds.
3. NSE calls `event.notification.setExtender` and alters the notification via extender -> Notification is shown, altered
4. NSE calls `event.preventDefault()` then starts a thread which calls `event.notification.display()` after 10 seconds  ->  Notification shows after 10 seconds.
5. NSE calls `event.preventDefault()` then starts a thread which calls `event.notification.display()` after 40 seconds ->  Notification does not show.
6. Foreground listener calls `event.preventDefault()` and returns immediately -> Notification does not show
7. Foreground listener calls `event.preventDefault()` then sleeps for 40 seconds before returning -> Notification shows after 30 seconds.
8. Foreground listener calls `event.preventDefault()` then starts a thread which calls `event.notification.display()` after 10 seconds  ->  Notification shows after 10 seconds.
9. Foreground listener calls `event.preventDefault()` then starts a thread which calls `event.notification.display()` after 40 seconds ->  Notification does not show.

I also confirmed the notification clicked listener and permission observer for appropriate behavior, though these were just name changes.

# Affected code checklist
   - [x] Notifications
      - [x] Display
      - [x] Open
      - [x] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [x] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Android-SDK/1765)

### CODE REVIEW

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.onesignal.notifications
+
+/**
+ * Implement this interface and provide an instance to [INotificationsManager.addForegroundLifecycleListener]
+ * in order to receive control during the lifecycle of the notification. This handler will *only* receive
+ * control when the application is in the foreground when the notification is received.
+ *
+ * @see [Foreground Notification Received Event | OneSignal Docs](https://documentation.onesignal.com/docs/sdk-notification-event-handlers#foreground-notification-received-event)
```
Comment (iAmWillShepherd) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt:8]: This should use our v11 docs which are here: https://dash.readme.com/project/onesignal/v11.0/docs/sdk-notification-event-handlers

Note that no one has provided a thorough review of these docs, and they may be wildly inaccurate or even incomplete.

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ */
+interface INotificationLifecycleListener {
+
+    /**
+     * Called when a notification is to be displayed to the user. This callback
+     * gives the implementor the ability to prevent the notification from displaying to the
+     * user entirely.
```
Comment (iAmWillShepherd) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt:15]:
```suggestion
     * user.
```

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+     *
+     * *Note:* this runs after the Notification Service Extension [INotificationServiceExtension]
+     * has been called (if one exists), which has the following differences:
```
Comment (iAmWillShepherd) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationLifecycleListener.kt:18]: > Note:* this runs after the Notification Service Extension [INotificationServiceExtension]
    has been called (if one exists), which has the following differences:

Revise the wording for clarity 👇🏽 

> If a Notification Service Extension [INotificationServiceExtension] exists, `onWillDisplay` will run after the extension is called...

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationReceivedEvent.kt [hunk 4]
```diff
@@ -1,25 +1,52 @@
...
+ * asynchronous processing to determine whether to display the notification).
+ *
+ * ```
+ * object : IRemoteNotificationReceivedHandler {
+ *   fun remoteNotificationReceived(event: IRemoteNotificationReceivedEvent) {
+ *     event.preventDefault()
+ *     thread {
+ *       // do something async
+ *
+ *       // optionally, change the display
```
Comment (iAmWillShepherd) [inline OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationReceivedEvent.kt:24]: Could you provide some guidance on how to "change  the display"?

**Review by iAmWillShepherd — CHANGES_REQUESTED** [review]
I had only a few nits that don't need to block this PR; however, a few points need to be addressed before I'm comfortable approving this. The note about the log should be applied to all files that make use of the format I specified although I only called it out once (this is my style in PR review to call something our once and avoid beating a dead 🐴 )

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationClickListener.kt [hunk 5]
```diff
@@ -4,15 +4,15 @@ package com.onesignal.notifications
  * An interface used to process a OneSignal notification the user just clicked on.
  * <br></br>
  * Set this during OneSignal init in
- * [INotificationsManager.setNotificationClickHandler]
+ * [INotificationsManager.addClickListener]
  * <br></br><br></br>
  * @see [NotificationOpenedHandler | OneSignal Docs](https://documentation.onesignal.com/docs/android-native-sdk.notificationopenedhandler)
```
[the remark this answers is not in the document]
Comment (brismithers) [reply OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationClickListener.kt:9]: This is a tricky one. I think I'd want to point to the "live" docs to cover any v11+ tweaks made?  I also can't find the appropriate page in v11 docs, here it is in the current docs: https://documentation.onesignal.com/docs/sdk-notification-event-handlers#notification-opened-event

File: OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationClickResult.kt [hunk 6]
```diff
@@ -1,17 +1,57 @@
...
-     * The notification that was opened by the user.
+     * The type of action the user took on this notification.
      */
-    val notification: INotification
+    val type: ActionType
+
+    /** When [type] is [ActionType.ActionButton], this will be the custom id of action taken.
+     *  See [Action Buttons | OneSignal Docs](https://documentation.onesignal.com/docs/action-buttons).
+     **/
+    val actionId: String?
```
[the remark this answers is not in the document]
Comment (brismithers) [reply OneSignalSDK/onesignal/core/src/main/java/com/onesignal/notifications/INotificationClickResult.kt:41]: When [type] is [ActionType.Opened] the `actionId` will be null.

File: Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/application/MainApplication.java [hunk 7]
```diff
@@ -76,23 +78,40 @@ public void inAppMessageClicked(@Nullable IInAppMessageClickResult result) {
...
+
+                //Prevent OneSignal from displaying the notification immediately on return. Spin
+                //up a new thread to mimic some asynchronous behavior, when the async behavior (which
+                //takes 2 seconds) completes, then the notification can be displayed.
+                event.preventDefault();
+                Runnable r = () -> {
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (brismithers) [reply Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/application/MainApplication.java:104]: Removed the printStackTrace altogether.  The purpose of this sleep is to mimic an async operations, if an `InterruptedException` occurs we can ignore it.

File: Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/notification/NotificationServiceExtension.java [hunk 8]
```diff
@@ -1,33 +1,27 @@
...
 import com.onesignal.sdktest.R;
 
-public class NotificationServiceExtension implements IRemoteNotificationReceivedHandler {
+public class NotificationServiceExtension implements INotificationServiceExtension {
 
    @Override
-   public void remoteNotificationReceived(Context context, INotificationReceivedEvent notificationReceivedEvent) {
-      Log.v("MainApplication", "IRemoteNotificationReceivedHandler fired!" + " with INotificationReceivedEvent: " + notificationReceivedEvent.toString());
+   public void onNotificationReceived(INotificationReceivedEvent event) {
+      Log.v("MainApplication", "IRemoteNotificationReceivedHandler fired!" + " with INotificationReceivedEvent: " + event.toString());
```
[the remark this answers is not in the document]
Comment (brismithers) [reply Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/notification/NotificationServiceExtension.java:15]: The pattern is more to tag log statements produced by the example app as "MainApplication" (vs the SDK which is tagged as "OneSignal").  The "format" of the log statement itself is context dependent, though it's common for verbose statement to provide the method name for easy identification.

I think the main confusion is using "MainApplication" as the tag, what do you think of me updating that to something like "SdkExampleApp"?

**Comment by brismithers** [comment]
@iAmWillShepherd pushed up changes. Please let me know if you have additional feedback!

File: Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/notification/NotificationServiceExtension.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (iAmWillShepherd) [reply Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/notification/NotificationServiceExtension.java]: I now see that this is a test. I don't think a change needs to be made.

**Review by iAmWillShepherd — APPROVED** [review]
LGTM 🚢

**Review by emawby — APPROVED** [review]

**Comment by iAmWillShepherd** [comment]
> Renamed all "handlers" to be either a "listener" or an "observer" depending on their purpose. All callbacks now have pattern "onXXXXXX" for each callback.

Could you share your thought process on which term to use? At the moment, it's not clear what you mean by their _purpose_.

**Comment by brismithers** [comment]
@iAmWillShepherd 

> Could you share your thought process on which term to use? At the moment, it's not clear what you mean by their purpose.

Realizing these terms are fuzzy in general, this is based on our team discussions to define as:
**Listener:** When you want to know when an event has occurred.
**Observer:** When you want to know when the state of an object (or property within an object) has changed.

**Pull request merged**